students/ArchWu/Session6/test_mailroom.py

```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python3
+
+"""
+unit tests for the mailroom program
+"""
+
+import mailroom as mailroom
+import os
+import pytest
+# so that it's there for the tests
+mailroom.donors = mailroom.get_donors()
+
+
+def test_list_donors():
+    listing = mailroom.list_donors()
+
+    # hard to test this throughly -- better not to hard code the entire
+    # thing. But check for a few aspects -- this will catch the likely
+    # errors
+    assert listing.startswith("Donor list:\n")
+    assert "Jeff Bezos" in listing
+    assert "William Gates III" in listing
+    assert len(listing.split('\n')) == 5
+
+
+def test_find_donor():
+    #pytest.skip("skipping this test")
+    """ checks a donor that is there, but with odd case and spaces"""
+    donor = mailroom.find_donor("jefF beZos ")
+
+    assert donor[0] == "Jeff Bezos"
+
+
+def test_find_donor_not():
+    "test one that's not there"
+    donor = mailroom.find_donor("Jeff Bzos")
+
+    assert donor is None
+
+
+def test_gen_letter():
+    """ test the donor letter """
+
+    # create a sample donor
+    donor = ("Fred Flintstone", [432.45, 65.45, 230.0])
+    letter = mailroom.gen_letter(donor)
+    # what to test? tricky!
+    assert letter.startswith("Dear Fred Flintstone")
+    assert letter.endswith("-The Team\n")
+    assert "donation of $230.00" in letter
+
+
+def test_add_donor():
+    """
+    adds a new donor
+
+    then tests that the donor is added, and that a donation is properly recorded.
+    """
+    name = "Fred Flintstone  "
+
+    donor = mailroom.add_donor(name)
+    donor[1].append(300)
+    assert donor[0] == "Fred Flintstone"
+    assert donor[1] == [300]
+    assert mailroom.find_donor(name) == donor
+
+
+def test_generate_donor_report():
+
+    report = mailroom.generate_donor_report()
+
+    print(report)  # printing so you can see it if it fails
+    # this is pretty tough to test
+    # these are not great, because they will fail if unimportant parts of the
+    # report are changed.
+    # but at least you know that codes working now.
+    assert report.startswith("Donor Name                | Total Given | Num Gifts | Average Gift")
+
+    assert "Jeff Bezos                  $    877.33           1   $     877.33" in report
```

Comment:
It may be easier to split this into multiple asserts so you dont need to worry about the whitespacing:
assert "Jeff Bezos" in report
assert "844.33" in report
etc